<a href="https://colab.research.google.com/github/NeSma237/GPT-2-Prompt-Engineering/blob/main/GPT_2_Prompt_Engineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [9]:
from transformers import GPT2LMHeadModel, GPT2Tokenizer
import torch

# Specify the model name
model_name = "gpt2"

# Load the tokenizer and model from Hugging Face
tokenizer = GPT2Tokenizer.from_pretrained(model_name)
model = GPT2LMHeadModel.from_pretrained(model_name)

# Use GPU if available, otherwise fallback to CPU
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model.to(device)


GPT2LMHeadModel(
  (transformer): GPT2Model(
    (wte): Embedding(50257, 768)
    (wpe): Embedding(1024, 768)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-11): 12 x GPT2Block(
        (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (attn): GPT2Attention(
          (c_attn): Conv1D(nf=2304, nx=768)
          (c_proj): Conv1D(nf=768, nx=768)
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): GPT2MLP(
          (c_fc): Conv1D(nf=3072, nx=768)
          (c_proj): Conv1D(nf=768, nx=3072)
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=768, out_features=50257, bias=False)
)

# Part 2  Design Your Prompts


---


1. Direct instruction:
"Write a motivational quote about overcoming fear."

2. Scenario-based:  
"Imagine you’re helping a friend who failed a test. Write something encouraging."

3. Persona-based:  
"As a wise monk, write a quote about inner strength."

4. Keyword-based:  
"Using the words 'growth', 'struggle', and 'hope', write something inspiring."

5. Conversational:  
"User: I feel like giving up. GPT-2: Here's a quote for you:"


In [10]:
prompts = {
    "Direct instruction": "Write a motivational quote about overcoming fear.",
    "Scenario-based": "Imagine you’re helping a friend who failed a test. Write something encouraging.",
    "Persona-based": "As a wise monk, write a quote about inner strength.",
    "Keyword-based": "Using the words 'growth', 'struggle', and 'hope', write something inspiring.",
    "Conversational": "User: I feel like giving up. GPT-2: Here's a quote for you:"
}


In [15]:
def generate_multiple_outputs(prompt, num_outputs=3, max_length=50):
    outputs = []
    input_ids = tokenizer.encode(prompt, return_tensors='pt').to(device)
    for _ in range(num_outputs):
       output = model.generate(
        input_ids,
        max_length=max_length,
        do_sample=True,
        temperature=0.8,
        top_k=50,
        top_p=0.95,
        num_return_sequences=1,
        pad_token_id=tokenizer.eos_token_id  # Set the padding token ID
)

       generated_text = tokenizer.decode(output[0], skip_special_tokens=True)
       outputs.append(generated_text)
    return outputs


In [16]:
for style, prompt in prompts.items():
    print(f"\n--- {style} Prompt ---")
    outputs = generate_multiple_outputs(prompt)
    for i, text in enumerate(outputs, 1):
        print(f"Output {i}:\n{text}\n")



--- Direct instruction Prompt ---
Output 1:
Write a motivational quote about overcoming fear. This may help you become more productive.

Try to make a few changes to your routine.

Try to keep your diet low, and at the same time, give yourself a break.



Output 2:
Write a motivational quote about overcoming fear.

6. Talk about how you learned to love yourself.

7. Show a lot of love.

8. Tell your kids that you love them.

9. Don't be

Output 3:
Write a motivational quote about overcoming fear. You can also add a note about how you will deal with the pain and fear from this problem.

2.

You'll begin with a short statement about how you feel.

What do


--- Scenario-based Prompt ---
Output 1:
Imagine you’re helping a friend who failed a test. Write something encouraging. And when you're not helping someone else, you'll probably find yourself wondering, "Why am I telling you this?"

5. Don't let your emotions

Output 2:
Imagine you’re helping a friend who failed a test. Write somethi

## Part 3 – Human-Written References

### 1. Direct Instruction Prompt
**Prompt:** Write a motivational quote about overcoming fear.  
**Human Reference:** “Everything you’ve ever wanted is on the other side of fear.” – George Addair  
**Source:** https://www.goodreads.com/quotes/2275

### 2. Scenario-Based Prompt
**Prompt:** Imagine you’re helping a friend who failed a test. Write something encouraging.  
**Human Reference:** “Failure is simply the opportunity to begin again, this time more intelligently.” – Henry Ford  
**Source:** https://www.brainyquote.com/quotes/henry_ford_131621

### 3. Persona-Based Prompt
**Prompt:** As a wise monk, write a quote about inner strength.  
**Human Reference:** “You have power over your mind — not outside events. Realize this, and you will find strength.” – Marcus Aurelius  
**Source:** https://www.goodreads.com/quotes/12478

### 4. Keyword-Based Prompt
**Prompt:** Using the words 'growth', 'struggle', and 'hope', write something inspiring.  
**Human Reference:** “Out of difficulties grow miracles.” – Jean de La Bruyère  
**Source:** https://www.goodreads.com/quotes/6867

### 5. Conversational Prompt
**Prompt:** User: I feel like giving up. GPT-2: Here's a quote for you:  
**Human Reference:** “When you feel like giving up, remember why you held on for so long in the first place.” – Unknown  
**Source:** https://www.keepinspiring.me/quotes-about-not-giving-up/


In [11]:
!pip install bert_score


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.1/61.1 kB 3.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 1.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 103.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 57.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 43.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 5.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 14.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 7.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 4.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 31.9 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstallin

In [17]:
#!pip install bert_score

from bert_score import score

human_references = {
    "Direct instruction": "Overcoming fear is the first step to success and growth.",
    "Scenario-based": "Failing is not the end; it's a chance to try again smarter and stronger.",
    "Persona-based": "True inner strength comes from understanding and accepting yourself.",
    "Keyword-based": "Growth often comes through struggle, but hope keeps us moving forward.",
    "Conversational": "Don't give up; every challenge is a step toward your better self."
}

all_outputs = {
    "Direct instruction": [
        "Write a motivational quote about overcoming fear. This may help you become more productive.\nTry to make a few changes to your routine.\nTry to keep your diet low, and at the same time, give yourself a break.",
        "Write a motivational quote about overcoming fear.\n6. Talk about how you learned to love yourself.\n7. Show a lot of love.\n8. Tell your kids that you love them.\n9. Don't be",
        "Write a motivational quote about overcoming fear. You can also add a note about how you will deal with the pain and fear from this problem.\n2.\nYou'll begin with a short statement about how you feel.\nWhat do"
    ],
    "Scenario-based": [
        "Imagine you’re helping a friend who failed a test. Write something encouraging. And when you're not helping someone else, you'll probably find yourself wondering, \"Why am I telling you this?\"\n5. Don't let your emotions",
        "Imagine you’re helping a friend who failed a test. Write something encouraging. Take a deep breath. Don't give up. Take a big hard look. If your friend is struggling to get through his test, go to your doctor.",
        "Imagine you’re helping a friend who failed a test. Write something encouraging. If someone makes you feel better about yourself, please think about it. If you feel depressed, want to help, or want to make friends, try something new."
    ],
    "Persona-based": [
        "As a wise monk, write a quote about inner strength. I would like to know if you feel the same way.\nSo, what can we do to increase our personal ability to do so?\n1. Make sure we understand our",
        "As a wise monk, write a quote about inner strength.\nIn the book, you will also learn to think as well as you think, and to become more like your body. The best part about learning to write a quote is that you",
        "As a wise monk, write a quote about inner strength. Then, you can tell me how to improve your life.\nWhat's the most difficult thing you have ever done?\nI'm not a bad person, but I'm not"
    ],
    "Keyword-based": [
        "Using the words 'growth', 'struggle', and 'hope', write something inspiring. The hope is to build something that will make it happen. The hope is to make it possible to overcome our difficulties and turn away from the struggle. This",
        "Using the words 'growth', 'struggle', and 'hope', write something inspiring. A successful and successful person needs to have a deep understanding of what they have just experienced.\nWhen you've experienced success, it's time to start",
        "Using the words 'growth', 'struggle', and 'hope', write something inspiring.\n\"The story of my life begins with the work I put into teaching my children about the world, and how they can learn to value it,"
    ],
    "Conversational": [
        "User: I feel like giving up. GPT-2: Here's a quote for you: \"If you want to participate in a group (and you want to do it on your own), we're going to give you that!\" You can",
        "User: I feel like giving up. GPT-2: Here's a quote for you:\n\"This is not about getting you back to the game.\"\n\"No, I'm not going to give up on it.\"",
        "User: I feel like giving up. GPT-2: Here's a quote for you:\n\"It's your right to be there as long as you want,\" said Ralston. \"But it's also your right to be"
    ]
}

for prompt_type, outputs in all_outputs.items():
    refs = [human_references[prompt_type]] * len(outputs)
    P, R, F1 = score(outputs, refs, lang="en", verbose=False)
    print(f"\n--- {prompt_type} ---")
    for i, (p, r, f1) in enumerate(zip(P, R, F1), 1):
        print(f"Output {i}: Precision={p:.4f}, Recall={r:.4f}, F1={f1:.4f}")


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.



--- Direct instruction ---
Output 1: Precision=0.8721, Recall=0.8985, F1=0.8851
Output 2: Precision=0.8618, Recall=0.8773, F1=0.8695
Output 3: Precision=0.8552, Recall=0.8903, F1=0.8724


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.



--- Scenario-based ---
Output 1: Precision=0.8393, Recall=0.8522, F1=0.8457
Output 2: Precision=0.8503, Recall=0.8600, F1=0.8551
Output 3: Precision=0.8486, Recall=0.8562, F1=0.8524


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.



--- Persona-based ---
Output 1: Precision=0.8402, Recall=0.8906, F1=0.8647
Output 2: Precision=0.8370, Recall=0.8850, F1=0.8603
Output 3: Precision=0.8302, Recall=0.8860, F1=0.8572


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.



--- Keyword-based ---
Output 1: Precision=0.8367, Recall=0.8688, F1=0.8525
Output 2: Precision=0.8323, Recall=0.8681, F1=0.8498
Output 3: Precision=0.8315, Recall=0.8737, F1=0.8521


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.



--- Conversational ---
Output 1: Precision=0.8481, Recall=0.8823, F1=0.8649
Output 2: Precision=0.8473, Recall=0.8847, F1=0.8656
Output 3: Precision=0.8285, Recall=0.8675, F1=0.8475


In [18]:
import pandas as pd
from bert_score import score

human_references = {
    "Direct instruction": "Overcoming fear is the first step to success and growth.",
    "Scenario-based": "Failing is not the end; it's a chance to try again smarter and stronger.",
    "Persona-based": "True inner strength comes from understanding and accepting yourself.",
    "Keyword-based": "Growth often comes through struggle, but hope keeps us moving forward.",
    "Conversational": "Don't give up; every challenge is a step toward your better self."
}

all_outputs = {
    "Direct instruction": [
        "Write a motivational quote about overcoming fear. This may help you become more productive.\nTry to make a few changes to your routine.\nTry to keep your diet low, and at the same time, give yourself a break.",
        "Write a motivational quote about overcoming fear.\n6. Talk about how you learned to love yourself.\n7. Show a lot of love.\n8. Tell your kids that you love them.\n9. Don't be",
        "Write a motivational quote about overcoming fear. You can also add a note about how you will deal with the pain and fear from this problem.\n2.\nYou'll begin with a short statement about how you feel.\nWhat do"
    ],
    "Scenario-based": [
        "Imagine you’re helping a friend who failed a test. Write something encouraging. And when you're not helping someone else, you'll probably find yourself wondering, \"Why am I telling you this?\"\n5. Don't let your emotions",
        "Imagine you’re helping a friend who failed a test. Write something encouraging. Take a deep breath. Don't give up. Take a big hard look. If your friend is struggling to get through his test, go to your doctor.",
        "Imagine you’re helping a friend who failed a test. Write something encouraging. If someone makes you feel better about yourself, please think about it. If you feel depressed, want to help, or want to make friends, try something new."
    ],
    "Persona-based": [
        "As a wise monk, write a quote about inner strength. I would like to know if you feel the same way.\nSo, what can we do to increase our personal ability to do so?\n1. Make sure we understand our",
        "As a wise monk, write a quote about inner strength.\nIn the book, you will also learn to think as well as you think, and to become more like your body. The best part about learning to write a quote is that you",
        "As a wise monk, write a quote about inner strength. Then, you can tell me how to improve your life.\nWhat's the most difficult thing you have ever done?\nI'm not a bad person, but I'm not"
    ],
    "Keyword-based": [
        "Using the words 'growth', 'struggle', and 'hope', write something inspiring. The hope is to build something that will make it happen. The hope is to make it possible to overcome our difficulties and turn away from the struggle. This",
        "Using the words 'growth', 'struggle', and 'hope', write something inspiring. A successful and successful person needs to have a deep understanding of what they have just experienced.\nWhen you've experienced success, it's time to start",
        "Using the words 'growth', 'struggle', and 'hope', write something inspiring.\n\"The story of my life begins with the work I put into teaching my children about the world, and how they can learn to value it,"
    ],
    "Conversational": [
        "User: I feel like giving up. GPT-2: Here's a quote for you: \"If you want to participate in a group (and you want to do it on your own), we're going to give you that!\" You can",
        "User: I feel like giving up. GPT-2: Here's a quote for you:\n\"This is not about getting you back to the game.\"\n\"No, I'm not going to give up on it.\"",
        "User: I feel like giving up. GPT-2: Here's a quote for you:\n\"It's your right to be there as long as you want,\" said Ralston. \"But it's also your right to be"
    ]
}

# Prepare data list for the DataFrame
data = []

for prompt_type, outputs in all_outputs.items():
    references = [human_references[prompt_type]] * len(outputs)
    P, R, F1 = score(outputs, references, lang="en", verbose=False)
    for i, f1 in enumerate(F1, start=1):
        data.append({
            "Prompt Type": prompt_type,
            "Output #": i,
            "BERTScore F1": f1.item()
        })

# Create DataFrame and display
df_results = pd.DataFrame(data)
print(df_results)

# Optionally save to CSV
df_results.to_csv("bert_score_results.csv", index=False)


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']
You sho

           Prompt Type  Output #  BERTScore F1
0   Direct instruction         1      0.885096
1   Direct instruction         2      0.869458
2   Direct instruction         3      0.872385
3       Scenario-based         1      0.845679
4       Scenario-based         2      0.855129
5       Scenario-based         3      0.852401
6        Persona-based         1      0.864670
7        Persona-based         2      0.860307
8        Persona-based         3      0.857188
9        Keyword-based         1      0.852462
10       Keyword-based         2      0.849810
11       Keyword-based         3      0.852110
12      Conversational         1      0.864858
13      Conversational         2      0.865580
14      Conversational         3      0.847516


**Analysis:**

The highest BERTScore was achieved by the first output of the Direct Instruction prompt (0.8851), indicating it has the greatest semantic similarity to the human-written motivational quote. Overall, all prompts generated outputs with relatively high semantic similarity scores (above 0.84), showing that GPT-2 can produce motivational text closely aligned with the reference across diverse prompt styles.